waiter/src/waiter/auth/oidc.clj

```diff
@@ -308,18 +309,41 @@
                  (= "." accept-redirect-auth)
                  (some #(= oidc-authority %) (str/split accept-redirect-auth #" ")))))))
 
+(defn count-challenge-cookies
+  "Returns the number of challenge cookies."
+  [request]
+  (let [cookie-header (get-in request [:headers "cookie"])
+        request-cookies (cond->> cookie-header
+                          (not (string? cookie-header)) (str/join ";"))]
+    (count
+      (filter #(str/starts-with? (str/trim %) oidc-challenge-cookie-prefix)
+              (str/split (str request-cookies) #";")))))
+
 (defn too-many-oidc-challenge-cookies?
   "Returns true if the request already contains too many OIDC challenge cookies."
   [request num-allowed]
-  (let [cookie-header (get-in request [:headers "cookie"])
-        request-cookies (cond->> cookie-header
-                          (not (string? cookie-header)) (str/join ";"))
-        num-challenge-cookies (count
-                                (filter #(str/starts-with? (str/trim %) oidc-challenge-cookie-prefix)
-                                        (str/split (str request-cookies) #";")))]
+  (let [num-challenge-cookies (count-challenge-cookies request)]
     (log/info "request has" num-challenge-cookies "oidc challenge cookies")
     (> num-challenge-cookies num-allowed)))
 
+(defn make-oidc-auth-too-many-cookies-response-updater
+  "Returns a response updater that rewrites 401 waiter responses to 302 redirects."
+  [request num-allowed]
+  (fn update-oidc-auth-too-many-cookies-response
+    [{:keys [waiter/auth-disabled? status] :as response}]
+    (let [num-challenge-cookies (count-challenge-cookies request)]
+      (cond-> response
+        (and auth-disabled?
+             (= status http-401-unauthorized)
+             (utils/waiter-generated-response? response))
```

Comment:
I don't understand what's going on here. We probably need a comment explaining why we need these checks.

Reply:
Discussed offline, updated with comments.